File: src/schema/application/organization_accounts.ts
import { boolean, index, pgTable, text } from 'drizzle-orm/pg-core';
import * as path from 'path';
import {
  fileRegex,
  system_fields,
} from '@dna-platform/crdt-lww-postgres/build/schema/system';
const filename = path.basename(__filename).replace(fileRegex, '');
import { organizations, external_contacts } from '..';
import { table as contacts } from './contacts';
import { table as organization_contacts } from './organization_contacts';
import { table as user_roles } from './user_roles';

const fields = {
  organization_contact_id: text().references(
    () => (organization_contacts as Record<string, any>).id,
  ),
  organization_id: text().references(() => organizations.id),
  contact_id: text().references(() => (contacts as Record<string, any>).id),
  email: text(),
  password: text(),
  account_id: text(),
  account_secret: text(),
  role_id: text().references(() => (user_roles as Record<string, any>).id),
  account_organization_id: text().references(
    () => (organizations as Record<string, any>).id,
  ),
  is_new_user: boolean().default(false),
  account_status: text(),
  external_contact_id: text().references(
    () => (external_contacts as Record<string, any>).id,
  ),
};

export const table = pgTable(
  filename,
  {
    ...system_fields,
    ...fields,
  },
  (table: Record<string, any>) => ({
    ...Object.keys({ ...system_fields, ...fields }).reduce((acc, field) => {
      const index_name = `organization_accounts_${field}_idx`;
      if (index_name.includes('_time')) return acc;
      return {
        ...acc,
        [index_name]: index(index_name).on(table[field]),
      };
    }, {}),
  }),
);

File: bin/store/src/database/schema/tables/organization_accounts.rs
use crate::define_table_schema;
use crate::schema::generator::diesel_schema_definition::{types::*, DieselTableDefinition};

pub struct OrganizationAccountsTable;

define_table_schema! {
    hypertable: false,
    fields: {
        system_fields!(),
        organization_contact_id: nullable(text()),
        organization_id: nullable(text()),
        contact_id: nullable(text()),
        email: nullable(text()),
        password: nullable(text()),
        account_id: nullable(text()),
        account_secret: nullable(text()),
        role_id: nullable(text()),
        account_organization_id: nullable(text()),
        is_new_user: nullable(boolean()), default: "false",
        account_status: nullable(text()),
        external_contact_id: nullable(text()),
        // Additional fields from original database schema
        image_url: nullable(text()),
        device_id: nullable(text()),
    },
    indexes: {
        system_indexes!("organization_accounts"),
        idx_organization_accounts_organization_contact_id: { columns: ["organization_contact_id"], unique: false, type: "btree" },
        idx_organization_accounts_contact_id: { columns: ["contact_id"], unique: false, type: "btree" },
        idx_organization_accounts_email: { columns: ["email"], unique: false, type: "btree" },
        idx_organization_accounts_password: { columns: ["password"], unique: false, type: "btree" },
        idx_organization_accounts_account_id: { columns: ["account_id"], unique: false, type: "btree" },
        idx_organization_accounts_account_secret: { columns: ["account_secret"], unique: false, type: "btree" },
        idx_organization_accounts_role_id: { columns: ["role_id"], unique: false, type: "btree" },
        idx_organization_accounts_account_organization_id: { columns: ["account_organization_id"], unique: false, type: "btree" },
        idx_organization_accounts_is_new_user: { columns: ["is_new_user"], unique: false, type: "btree" },
        idx_organization_accounts_account_status: { columns: ["account_status"], unique: false, type: "btree" },
        idx_organization_accounts_external_contact_id: { columns: ["external_contact_id"], unique: false, type: "btree" },
        idx_organization_accounts_image_url: { columns: ["image_url"], unique: false, type: "btree" },
        idx_organization_accounts_device_id: { columns: ["device_id"], unique: false, type: "btree" },
    },
    foreign_keys: {
        system_foreign_keys!("organization_accounts"),
        fk_organization_accounts_organization_contact_id: { columns: ["organization_contact_id"], foreign_table: "organization_contacts", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
        fk_organization_accounts_organization_id: { columns: ["organization_id"], foreign_table: "organizations", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
        fk_organization_accounts_contact_id: { columns: ["contact_id"], foreign_table: "contacts", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
        fk_organization_accounts_role_id: { columns: ["role_id"], foreign_table: "user_roles", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
        fk_organization_accounts_account_organization_id: { columns: ["account_organization_id"], foreign_table: "organizations", foreign_columns: ["id"], on_delete: "no action", on_update: "no action" },
        fk_organization_accounts_device_id: { columns: ["device_id"], foreign_table: "devices", foreign_columns: ["id"], on_delete: "set null", on_update: "no action" },
    }
}
